### PR TITLE
Make `clone` abstract again

### PR DESCRIPTION
**Pull request by nechaido**
* Rewrite `clone` so it properly copies a prototype.
  This way we will need only one method instead of 3
* Add optional clone depth parameter

### CODE REVIEW

**Comment by nechaido** [comment]
@tshemsedinov can`t get it, both what?

**Comment by tshemsedinov** [comment]
* Clone data structures (array and object) with old implementation `api.common.clone`
* Clone with class/proto instantiation as you proposed, but need different name for it `api.common.objClone` or deepClone, but I have no good name for it now

After that we may compare: copy, clone, deepClone, extend, Object.assign, etc.

**Comment by tshemsedinov** [comment]
`clone` has expected behavior (not only in Impress, it's from underscore and such libs) and it should be less functional but performant

**Comment by aqrln** [comment]
While this implementation is much shorter, prettier and covers more usecases, I have admit that it is, maybe, too generic and we'll still have to have specialized functions too, since this PR brings a significant performance regression (at least, when cloning arrays).

I though it could be improved by making arrays a special case, but it didn't give a very big boost.

<details>
  <summary>Patch to @nechaido's implementation</summary>
  <p>

```diff
diff --git a/lib/data.js b/lib/data.js
index 710617a..6cb1fbb 100644
--- a/lib/data.js
+++ b/lib/data.js
@@ -32,6 +32,16 @@ api.common.clone = (
     return obj;
   }
 
+  if (Array.isArray(obj)) {
+    const size = obj.length;
+    const array = new Array(size);
+    let i;
+    for (i = 0; i < size; i++) {
+      array[i] = deepClone(obj[i], depth && depth - 1);
+    }
+    return array;
+  }
+
   const clone = new obj.constructor();
   Object.keys(obj).forEach(key => {
     clone[key] = api.common.clone(obj[key], depth ? depth - 1 : undefined);
```

  </p>
</details>
<br>

Though the original `cloneArray` function can be improved too.

<details>
  <summary>Patch to @tshemsedinov's implementation</summary>
  <p>

```diff
diff --git a/lib/data.js b/lib/data.js
index 5797fe9..23ea0dd 100644
--- a/lib/data.js
+++ b/lib/data.js
@@ -50,18 +50,16 @@ function cloneObject(obj) {
 }
 
 function cloneArray(arr) {
-  const result = [];
   const len = arr.length;
+  const result = new Array(len);
   let i, val, type;
   for (i = 0; i < len; i++) {
     val = arr[i];
     type = typeof(val);
     if (type === 'object') {
-      result.push(
-        Array.isArray(val) ? cloneArray(val) : cloneObject(val)
-      );
+      result[i] = Array.isArray(val) ? cloneArray(val) : cloneObject(val);
     } else {
-      result.push(val);
+      result[i] = val;
     }
   }
   return result;
```

  </p>
</details>
<br>

Benchmark results:

```
tshemsedinov's cloneArray x 3,317,870 ops/sec ±1.25% (85 runs sampled)
nechaido's clone x 137,082 ops/sec ±0.73% (90 runs sampled)
patched tshemsedinov's cloneArray x 3,516,861 ops/sec ±1.31% (91 runs sampled)
patched nechaido's clone x 162,357 ops/sec ±0.88% (91 runs sampled)
Fastest is patched tshemsedinov's cloneArray
```

[Benchmark source code](https://gist.github.com/aqrln/586be14f0132e4e449f375536582e06f)

**Comment by aqrln** [comment]
Sorry, my patched functions called untouched ones recursively :) After fixing the mistake, the results are much more interesting:

```
tshemsedinov's cloneArray x 3,467,874 ops/sec ±0.96% (90 runs sampled)
nechaido's clone x 136,195 ops/sec ±0.81% (93 runs sampled)
patched tshemsedinov's cloneArray x 5,448,285 ops/sec ±1.32% (90 runs sampled)
patched nechaido's clone x 2,971,660 ops/sec ±1.20% (91 runs sampled)
Fastest is patched tshemsedinov's cloneArray
```

**Comment by nechaido** [comment]
Removed clone depth parameter in order to make function faster, here are results of [benchmark](https://gist.github.com/nechaido/8c8978d12a2771d98952b6bd892faef4#file-metarhia_console_clone_benchmark-js)
```
Pure arrays:
tshemsedinov`s x 1,008,097 ops/sec ±4.13% (81 runs sampled)
nechaido`s x 58,271 ops/sec ±5.17% (81 runs sampled)
patched tshemsedinov's x 1,858,912 ops/sec ±2.97% (85 runs sampled)
patched nechaido's x 2,120,233 ops/sec ±2.90% (83 runs sampled)
Fastest is patched nechaido's
Objects:
tshemsedinov`s x 96,704 ops/sec ±0.84% (88 runs sampled)
nechaido`s x 12,552 ops/sec ±5.06% (79 runs sampled)
patched tshemsedinov's x 80,500 ops/sec ±4.02% (72 runs sampled)
patched nechaido's x 99,229 ops/sec ±1.09% (91 runs sampled)
Fastest is patched nechaido's
```

**Comment by nechaido** [comment]
@tshemsedinov @aqrln i`d like you to rereview this

File: lib/data.js [hunk 1]
```diff
@@ -24,47 +24,40 @@ api.common.copy = (
...
+  const size = arr.length;
+  const array = new Array(size);
+  let i, val;
+  for (i = 0; i < size; i++) {
+    val = arr[i];
+    if (typeof val !== 'object' || val === null) {
+      array[i] = val;
     } else {
-      result[key] = val;
+      array[i] = Array.isArray(val) ? cloneArray(val) : cloneObject(val);
```
Comment (aqrln) [inline lib/data.js]: Why not use `api.common.clone()` here? It already contains this type checking logic.

File: lib/data.js [hunk 2]
```diff
@@ -24,47 +24,40 @@ api.common.copy = (
...
+  const keys = Object.keys(obj);
+  const size = keys.length;
+  let i, val;
+  for (i = 0; i < size; i++) {
+    val = obj[keys[i]];
+    if (typeof val !== 'object' || val === null) {
+      clone[keys[i]] = val;
     } else {
-      result.push(val);
+      clone[keys[i]] = Array.isArray(val) ? cloneArray(val) : cloneObject(val);
```
Comment (aqrln) [inline lib/data.js]: Same as https://github.com/metarhia/Common/pull/1/files#r104704000

File: lib/data.js (hunk 1 above)
Comment (nechaido) [inline lib/data.js]: If i reuse the `clone` function it losses for about 15% of performance

File: lib/data.js [hunk 3]
```diff
@@ -24,47 +24,40 @@ api.common.copy = (
...
-  let i, val, type;
-  for (i = 0; i < len; i++) {
-    val = arr[i];
-    type = typeof(val);
-    if (type === 'object') {
-      result.push(
-        Array.isArray(val) ? cloneArray(val) : cloneObject(val)
-      );
+function cloneObject(obj) {
+  const clone = obj.constructor();
```
Comment (nechaido) [inline lib/data.js]: My fault, i`ll fix it now

Comment (nechaido) [inline lib/data.js]: fixed

File: lib/data.js [hunk 4]
```diff
@@ -24,47 +24,40 @@ api.common.copy = (
...
-        Array.isArray(val) ? cloneArray(val) : cloneObject(val)
-      );
+function cloneObject(obj) {
+  const clone = new obj.constructor();
+  const keys = Object.keys(obj);
+  const size = keys.length;
+  let i, val;
+  for (i = 0; i < size; i++) {
+    val = obj[keys[i]];
+    if (typeof val !== 'object' || val === null) {
```
Comment (tshemsedinov) [inline lib/data.js]: We use `typeof(val)`

**Comment by nechaido** [comment]
@tshemsedinov what will be the difference between Universal clone and Clone object, also - why does Universal clone `api.common.clone()` and Clone data structures `api.common.clone()` have simmilar names?

**Comment by tshemsedinov** [comment]
Sorry, that's mistype, I mean clone data structures will be `api.common.data.clone()` and `api.common.clone()` will select `api.common.data.clone()` or `api.common.object.clone()` it would be great, if possible, need to try.

**Comment by nechaido** [comment]
@tshemsedinov Do i get it right?
* `copy` - deep Clone of everything, that handles cyclic refs
* `clone` - points to `object.clone` or `data.clone` we`ll need to decide which
* `object.clone` - deep Clone with proper prototype and cyclic refs handling
* `data.clone` - deep clone of data only structures with no prorotype or cyclic handling?

If that's right - what is the difference between `copy` and `object.clone`?

**Comment by nechaido** [comment]
@metarhia/inquisitori i`d like you to re review this PR

File: lib/data.js [hunk 5]
```diff
@@ -24,47 +24,97 @@ api.common.copy = (
   return result;
 */
 
-api.common.clone = (
-  // Clone dataset (clone objects to new array)
-  ds
-) => cloneArray(ds);
+api.common.clone = (obj) => api.common.data.clone(obj);
```
Comment (tshemsedinov) [inline lib/data.js]: `api.common.clone = api.common.data.clone;`

File: lib/data.js [hunk 6]
```diff
@@ -24,47 +24,97 @@ api.common.copy = (
...
+
+if (!api.common.data) api.common.data = {};
```
Comment (tshemsedinov) [inline lib/data.js]: `api.common.data = api.common.data || {};` but here we know that no `api.common.data` defined, so we can just use: `api.common.data = {};`

File: lib/data.js [hunk 7]
```diff
@@ -24,47 +24,97 @@ api.common.copy = (
...
-  return result;
+  return clone;
 }
 
-function cloneArray(arr) {
-  const result = [];
-  const len = arr.length;
-  let i, val, type;
-  for (i = 0; i < len; i++) {
+if (!api.common.object) api.common.object = {};
```
Comment (tshemsedinov) [inline lib/data.js]: `api.common.object = api.common.object || {};` or `api.common.object = {};`

**Comment by nechaido** [comment]
@tshemsedinov fixed

File: lib/data.js [hunk 8]
```diff
@@ -24,47 +24,97 @@ api.common.copy = (
   return result;
 */
 
-api.common.clone = (
-  // Clone dataset (clone objects to new array)
-  ds
-) => cloneArray(ds);
+api.common.clone = api.common.data.clone;
```
Comment (tshemsedinov) [inline lib/data.js]: `api.common.data.clone` is not defined yet

File: lib/data.js [hunk 9]
```diff
@@ -24,47 +24,97 @@ api.common.copy = (
...
     } else {
-      result.push(val);
+      array[i] = cloneObjObj(val, references);
     }
   }
-  return result;
+  return array;
+}
+
+function cloneObjObj(obj, references) {
```
Comment (nechaido) [inline lib/data.js]: It`s for cloning objects with prototypes and circular refs

**Comment by nechaido** [comment]
@tshemsedinov there are already 2 instances of `clone` for array and object, but they are not exposed to user

**Comment by nechaido** [comment]
@tshemsedinov done.
* `api.common.clone` - for data
* `api.common.duplicate` - for objects with prototypes and circular references

**Comment by tshemsedinov** [comment]
Landed in 5e7d7d30a78f89a0e46e01b7af510b8404a7d28a

**Comment by tshemsedinov** [comment]
@nechaido please see minor optimizations here: 840a40ed249964cf2465b355d9efb0ff6286fb55

**Comment by nechaido** [comment]
@tshemsedinov you are right, it's faster, anyway will nedd to aply it both for `cloneObj` and `duplicateObj`:
```
original x 113,730 ops/sec ±0.31% (1087 runs sampled)
patched x 114,777 ops/sec ±0.27% (1088 runs sampled)
Fastest is patched
```

**Comment by nechaido** [comment]
patched f5764555d52d0afda06cea4c4c369ed632e05e6e

**Comment by tshemsedinov** [comment]
Landed in 8c4713d

**Pull request closed without merging**